File: src/FastTest.py
import pandas as pd 
import numpy as np

from Strategy import Strategy
from Exchange import Exchange
from Broker import Broker
from Asset import Asset
from Order import *

class FastTest():
    def __init__(self, exchange : Exchange, broker : Broker, **kwargs) -> None:
        self.exchange = exchange
        self.broker = broker
        self.strategy = kwargs.get("strategy")
        self.exchange.build()
        self.portfolio_value_history = []
        self.benchmark_portfolio_value_history = []
        self.benchmark = False

    def reset(self):
        self.exchange.reset()
        self.broker.reset()
        self.portfolio_value_history = []
        self.benchmark_portfolio_value_history = []

    def register_strategy(self, strategy : Strategy):
        self.strategy = strategy
        self.strategy.strategy_id = "user"

    def register_benchmark(self, strategy : Strategy):
        self.benchmark_strategy = strategy
        self.benchmark_strategy.strategy_id = "benchmark"
        self.benchmark_broker = strategy.broker
        self.benchmark = True

    def analyze_strategy_on_next(self):
        self.portfolio_value_history.append(self.broker.net_liquidation_value)
        if self.benchmark: self.benchmark_portfolio_value_history.append(self.benchmark_strategy.broker.net_liquidation_value)

    def analyze_strategy_on_finish(self):
        try:
            min_warmup = min([asset.warmup for asset in self.exchange.market.values()])
        except:
            min_warmup = 0
        self.portfolio_value_history = pd.DataFrame(
            data = self.portfolio_value_history,
            index = self.exchange.datetime_index[min_warmup:],
            columns = ["net_liquidation_value"]
        )
        if self.benchmark: self.portfolio_value_history["benchmark"] = self.benchmark_portfolio_value_history

        for asset_analysis in self.broker.strategy_analysis.values():
            number_trades = (asset_analysis["wins"] + asset_analysis["losses"])
            if number_trades == 0: continue
            asset_analysis["number_trades"] = number_trades
            asset_analysis["win_rate"] = asset_analysis["wins"] / number_trades
            asset_analysis["average_time_in_market"] = asset_analysis["time_in_market"] / number_trades
            asset_analysis["average_position_size"] = asset_analysis["total_units"] / number_trades
            asset_analysis["average_pl"] = asset_analysis["pl"] / number_trades

    def run(self):  
        #clear any previous testing information
        self.reset()

        #intilze broker
        self.broker.build()
        if self.benchmark: self.benchmark_strategy.broker.build()

        while self.exchange.next():
            
            #check to see if any assets are available (allows for warmup)
            if len(self.exchange.market_view.keys()) == 1:
                continue
            
            #evaluate collateral held by broker
            self.broker.evaluate_collateral()

            #allow exchange to process open orders from previous steps
            filled_orders = self.exchange.process_orders(strategy_id = self.strategy.strategy_id)

            #allow broker to process orders that have been filled
            self.broker.process_filled_orders(filled_orders) 

            #allow strategy to place orders based on current market view. Orders processed next market open
            orders = self.strategy.next()

            #place the orders to the broker who routes them to the exchange
            self.broker.place_orders(orders,strategy_id = self.strategy.strategy_id)
            
            #value the portfolio
            self.broker.evaluate_portfolio()
            
            if self.benchmark:
                if self.benchmark_broker.cheat_on_close: 
                    orders_benchmark = self.benchmark_strategy.next()
                    self.benchmark_broker.place_orders(orders_benchmark,strategy_id = self.benchmark_strategy.strategy_id)
                    filled_orders = self.exchange.process_orders(strategy_id = self.benchmark_strategy.strategy_id, cheat_on_close = True)
                    self.benchmark_broker.process_filled_orders(filled_orders) 
                else:
                    filled_orders = self.exchange.process_orders(strategy_id = self.benchmark_strategy.strategy_id)
                    self.benchmark_broker.process_filled_orders(filled_orders)
                    orders_benchmark = self.benchmark_strategy.next()
                    self.benchmark_broker.place_orders(orders_benchmark,strategy_id = self.benchmark_strategy.strategy_id)

                orders_benchmark = self.benchmark_strategy.next()
                
                self.benchmark_strategy.broker.evaluate_portfolio()

            #analyze portfolio stats
            self.analyze_strategy_on_next()

        """End of backtest"""

        #clear all orders currently open on the exchange
        self.broker.clear_orders()
        
        #close all open positions at last close price for each asst
        self.broker.clear_portfolio()
        
        #provide a final valuation of the portfolio
        self.broker.evaluate_portfolio()
        
        if self.benchmark:
            self.benchmark_strategy.broker.clear_orders()
            self.benchmark_strategy.broker.clear_portfolio()
            self.benchmark_strategy.broker.evaluate_portfolio()

        #analyize strategy statistics
        self.analyze_strategy_on_finish()


if __name__ == "__main__":
    exchange = Exchange()
    broker = Broker(exchange=exchange)
    strategy = Strategy(exchange=exchange,broker=broker)

    source_type = "csv"
    datetime_format = "%Y-%m-%d"
    datetime_column = "DATE"

    tickers = ["AAL","AAP"]

    for ticker in tickers:
        csv_path = r"C:\Users\bktor\Desktop\Python\FastTest\data\{}.csv".format(ticker)
        name = ticker
        exchange.register_asset(Asset(
                ticker,
                source_type = source_type,
                csv_path = csv_path,
                datetime_format = datetime_format,
                datetime_column = datetime_column
            )
        )
    
    fast_test = FastTest(
        exchange=exchange,
        broker=broker,
        strategy=strategy
        )
    
    fast_test.run()

File: src/c++/FastTest.cpp
#include "pch.h"
#include "Order.h"
#include "Position.h"
#include "Asset.h"
#include "Exchange.h"
#include "Broker.h"
#include "FastTest.h"

__FastTest::__FastTest(__Exchange &exchangeObj, __Broker &brokerObj, Strategy *StrategyObj, bool logging) :
	__exchange(exchangeObj), broker(brokerObj), strategy(StrategyObj){
	this->logging = logging;
}
__FastTest::__FastTest(__Exchange &exchangeObj, __Broker &brokerObj, bool logging) :
	__exchange(exchangeObj), broker(brokerObj) {
	this->logging = logging;
}
void __FastTest::reset() {
	this->broker.reset();
	this->__exchange.reset();
	this->cash_history.clear();
	this->nlv_history.clear();
	this->filled_orders.clear();
	this->canceled_orders.clear();
}
void __FastTest::analyze_step() {
	this->cash_history.push_back(this->broker.cash);
	this->nlv_history.push_back(this->broker.net_liquidation_value);
}
void __FastTest::run() {
	if (this->logging) { printf("RUNNING FASTEST\n"); }
	this->__exchange.logging = this->logging;
	this->broker.logging = this->logging;
	this->reset();

	while (this->__exchange._get_market_view()) {
		//allow exchange to process open orders from previous steps
		if (!this->__exchange.orders.empty()) {
			filled_orders = this->__exchange.process_orders();
			this->broker.process_filled_orders(std::move(filled_orders));
		}
		//allow strategy to place orders
		this->strategy->next();

		//evaluate the portfolio
		this->broker.evaluate_portfolio();

		//evaluate strategy portfolio
		this->analyze_step();

		//allow the exchange to clean up assets that are done streaming
		canceled_orders = this->__exchange.clean_up_market();
		if (!canceled_orders.empty()) {
			this->broker.log_canceled_orders(std::move(canceled_orders));
		}

		//allow exchange to process cheat on close orders
		if (!this->__exchange.orders.empty()) {
			filled_orders = this->__exchange.process_orders(true);
			this->broker.process_filled_orders(std::move(filled_orders));
		}
		this->step_count++;
		if (this->logging) { printf("%i\n", step_count); }
	}
}
void * CreateFastTestPtr(void *exchange_ptr, void *broker_ptr, bool logging) {
	__Exchange *__exchange_ref = static_cast<__Exchange *>(exchange_ptr);
	__Broker *__broker_ref = static_cast<__Broker *>(broker_ptr);
	__FastTest* fastTest_ptr = new __FastTest(*__exchange_ref,*__broker_ref, logging);
	fastTest_ptr->__exchange.logging = logging;
	fastTest_ptr->broker.logging = logging;
	return fastTest_ptr;
}
void DeleteFastTestPtr(void *fastTest_ptr){
	delete fastTest_ptr;
}
void reset(void *fastTest_ptr) {
	__FastTest *__fastTest_ref = static_cast<__FastTest *>(fastTest_ptr);
	__fastTest_ref->reset();
}
bool forward_pass(void *fastTest_ptr){
	__FastTest *__fastTest_ref = static_cast<__FastTest *>(fastTest_ptr);
	
	if (!__fastTest_ref->__exchange._get_market_view()) { return false; }
	//allow exchange to process open orders from previous steps
	if (!__fastTest_ref->__exchange.orders.empty()) {
		__fastTest_ref->filled_orders = __fastTest_ref->__exchange.process_orders();
		__fastTest_ref->broker.process_filled_orders(std::move(__fastTest_ref->filled_orders));
	}
	return true;
}
void backward_pass(void * fastTest_ptr) {
	
	__FastTest *__fastTest_ref = static_cast<__FastTest *>(fastTest_ptr);
	//evaluate the portfolio
	__fastTest_ref->broker.evaluate_portfolio();
	//evaluate strategy portfolio
	__fastTest_ref->analyze_step();
	//allow the exchange to clean up assets that are done streaming
	__fastTest_ref->canceled_orders = __fastTest_ref->__exchange.clean_up_market();
	if (!__fastTest_ref->canceled_orders.empty()) {
		__fastTest_ref->broker.log_canceled_orders(std::move(__fastTest_ref->canceled_orders));
	}
	//allow exchange to process cheat on close orders
	if (!__fastTest_ref->__exchange.orders.empty()) {
		__fastTest_ref->filled_orders = __fastTest_ref->__exchange.process_orders(true);
		__fastTest_ref->broker.process_filled_orders(std::move(__fastTest_ref->filled_orders));
	}
	__fastTest_ref->step_count++;
}